### PR TITLE
feat(dlm):support breakpoint recovery

### PR DESCRIPTION
**Pull request by guowl3**
#### What type of PR is this?
type-feature

#### What this PR does / why we need it:

DLM support breakpoint recovery.
#### Which issue(s) this PR fixes:

Fixes #

#### Additional documentation e.g., usage docs, etc.:

```docs

```

### CODE REVIEW

File: server/odc-migrate/src/main/resources/migrate/common/V_4_2_2_4__add_dlm_task_meta.sql [hunk 1]
```diff
@@ -0,0 +1,31 @@
+CREATE TABLE IF NOT EXISTS `dlm_task_generator` (
+`id` bigint(20) NOT NULL AUTO_INCREMENT,
+`create_time` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP COMMENT '创建时间',
```
Comment (yhilmare) [inline server/odc-migrate/src/main/resources/migrate/common/V_4_2_2_4__add_dlm_task_meta.sql:3]: you may use english as column or table comment, just like what we did before

File: server/odc-migrate/src/main/resources/migrate/common/V_4_2_2_4__add_dlm_task_meta.sql [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+`job_id` varchar(120) NOT NULL,
+`processed_data_size` bigint(20) NOT NULL,
+`processed_row_count` bigint(20) NOT NULL,
+`status` varchar(64) NOT NULL,
+`type` varchar(32) NOT NULL,
+`primary_key_save_point` varchar(512) DEFAULT NULL,
+`partition_save_point` varchar(512) DEFAULT NULL,
+`task_count` bigint(20) NOT NULL DEFAULT '0',
+PRIMARY KEY (`id`),
+KEY `pk_generatora_id` (`generator_id`)
```
Comment (yhilmare) [inline server/odc-migrate/src/main/resources/migrate/common/V_4_2_2_4__add_dlm_task_meta.sql:15]: key's name is illegal, may 'pk_${table_name}_${ref_col_name}'

File: server/odc-migrate/src/main/resources/migrate/common/V_4_2_2_4__add_dlm_task_meta.sql [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+`generator_id` varchar(120) NOT NULL,
+`job_id` varchar(120) NOT NULL,
+`processed_data_size` bigint(20) NOT NULL,
+`processed_row_count` bigint(20) NOT NULL,
+`status` varchar(64) NOT NULL,
+`type` varchar(32) NOT NULL,
+`primary_key_save_point` varchar(512) DEFAULT NULL,
+`partition_save_point` varchar(512) DEFAULT NULL,
+`task_count` bigint(20) NOT NULL DEFAULT '0',
+PRIMARY KEY (`id`),
```
Comment (yhilmare) [inline server/odc-migrate/src/main/resources/migrate/common/V_4_2_2_4__add_dlm_task_meta.sql:14]: do not use anonymous pk, use `constraint pk_ dlm_task_generator_id primary key(`id`)` instead

File: server/odc-migrate/src/main/resources/migrate/common/V_4_2_2_4__add_dlm_task_meta.sql [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+KEY `pk_generatora_id` (`generator_id`)
+);
+CREATE TABLE `dlm_task_unit` (
+`id` bigint(20) NOT NULL AUTO_INCREMENT,
+`create_time` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP COMMENT '创建时间',
+`update_time` datetime DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP COMMENT '修改时间',
+`task_index` bigint(20) NOT NULL,
```
Comment (yhilmare) [inline server/odc-migrate/src/main/resources/migrate/common/V_4_2_2_4__add_dlm_task_meta.sql:21]: leak of column comment

File: server/odc-migrate/src/main/resources/migrate/common/V_4_2_2_4__add_dlm_task_meta.sql [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+`job_id` varchar(120) NOT NULL,
+`generator_id` varchar(120) NOT NULL,
+`status` varchar(64) NOT NULL,
+`lower_bound_primary_key` varchar(512) DEFAULT NULL,
+`upper_bound_primary_key` varchar(512) DEFAULT NULL,
+`primary_key_cursor` varchar(512) DEFAULT NULL,
+`partition_name` varchar(512) DEFAULT NULL,
+PRIMARY KEY (`id`),
```
Comment (yhilmare) [inline server/odc-migrate/src/main/resources/migrate/common/V_4_2_2_4__add_dlm_task_meta.sql:29]: same problem as above

File: server/odc-migrate/src/main/resources/migrate/common/V_4_2_2_4__add_dlm_task_meta.sql [hunk 6]
```diff
@@ -0,0 +1,31 @@
+CREATE TABLE IF NOT EXISTS `dlm_task_generator` (
+`id` bigint(20) NOT NULL AUTO_INCREMENT,
+`create_time` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP COMMENT '创建时间',
+`update_time` datetime DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP COMMENT '修改时间',
+`generator_id` varchar(120) NOT NULL,
```
Comment (yhilmare) [inline server/odc-migrate/src/main/resources/migrate/common/V_4_2_2_4__add_dlm_task_meta.sql:5]: the column would better get a comment

File: server/odc-service/src/main/java/com/oceanbase/odc/metadb/dlm/TaskGeneratorRepository.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+/**
+ * @Author：tinker
+ * @Date: 2023/8/10 14:48
+ * @Descripition:
+ */
+public interface TaskGeneratorRepository extends OdcJpaRepository<TaskGeneratorEntity, Long> {
+
+    Optional<TaskGeneratorEntity> findByJobId(String jobId);
+
+    Optional<TaskGeneratorEntity> findByGeneratorId(String generatorId);
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/metadb/dlm/TaskGeneratorRepository.java:31]: how can you guarantee that you will get a singleton object queried by generatorId? I don't see any constraints on it

File: server/odc-service/src/main/java/com/oceanbase/odc/metadb/dlm/TaskGeneratorRepository.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+import com.oceanbase.odc.config.jpa.OdcJpaRepository;
+
+/**
+ * @Author：tinker
+ * @Date: 2023/8/10 14:48
+ * @Descripition:
+ */
+public interface TaskGeneratorRepository extends OdcJpaRepository<TaskGeneratorEntity, Long> {
+
+    Optional<TaskGeneratorEntity> findByJobId(String jobId);
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/metadb/dlm/TaskGeneratorRepository.java:29]: same problem as follow

File: server/odc-service/src/main/java/com/oceanbase/odc/metadb/dlm/TaskUnitRepository.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+/**
+ * @Author：tinker
+ * @Date: 2023/8/10 15:20
+ * @Descripition:
+ */
+public interface TaskUnitRepository extends OdcJpaRepository<TaskUnitEntity, Long> {
+
+    List<TaskUnitEntity> findByGeneratorId(@Param("generatorId") String generatorId);
+
+    Optional<TaskUnitEntity> findByJobIdAndGeneratorIdAndTaskIndex(String jobId, String generatorId, Long taskIndex);
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/metadb/dlm/TaskUnitRepository.java:35]: how can you ensure that there will be a singleton object queried by these conditions?

File: server/odc-service/src/main/java/com/oceanbase/odc/metadb/dlm/TaskUnitRepository.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+    @Query(value = "select count(1) from dlm_task_unit where job_id=:jobId and (status != 'SUCCESS' or primary_key_cursor is null)",
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/metadb/dlm/TaskUnitRepository.java:37]: magic value, why 'SUCCESS'? could it be changed in the future?

Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/metadb/dlm/TaskUnitRepository.java:37]: is `primary_key_cursor is null` a valid grammar in h2?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/dlm/TaskGeneratorMapper.java [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+/**
+ * @Author：tinker
+ * @Date: 2023/8/10 15:47
+ * @Descripition:
+ */
+public class TaskGeneratorMapper {
+
+
+    public static TaskGenerator entityToModel(TaskGeneratorEntity entity) {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/dlm/TaskGeneratorMapper.java:34]: there is no need to do this, you may use `@Mapper`, just like what `ConnectionMapper` did

File: server/odc-service/src/main/java/com/oceanbase/odc/service/dlm/TaskUnitMapper.java [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+
+/**
+ * @Author：tinker
+ * @Date: 2023/8/10 17:26
+ * @Descripition:
+ */
+public class TaskUnitMapper {
+
+
+    public static TaskMeta entityToModel(TaskUnitEntity entity) {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/dlm/TaskUnitMapper.java:32]: as above

File: server/odc-service/src/main/java/com/oceanbase/odc/metadb/dlm/TaskGeneratorRepository.java (hunk 7 above)
Comment (guowl3) [reply server/odc-service/src/main/java/com/oceanbase/odc/metadb/dlm/TaskGeneratorRepository.java:28]: This method is unused and has been deleted.

File: server/odc-service/src/main/java/com/oceanbase/odc/metadb/dlm/TaskUnitRepository.java (hunk 9 above)
Comment (guowl3) [reply server/odc-service/src/main/java/com/oceanbase/odc/metadb/dlm/TaskUnitRepository.java:35]: I have added a joint unique index to this table

File: server/odc-service/src/main/java/com/oceanbase/odc/metadb/dlm/TaskUnitRepository.java (hunk 10 above)
Comment (guowl3) [reply server/odc-service/src/main/java/com/oceanbase/odc/metadb/dlm/TaskUnitRepository.java:37]: This is the SQL of the original DLM. This time we will keep it consistent and will reconstruct it later.

Comment (guowl3) [reply server/odc-service/src/main/java/com/oceanbase/odc/metadb/dlm/TaskUnitRepository.java:37]: DLM is unsupported in client. We will not modify it this time and will reconstruct it later.

File: server/odc-migrate/src/main/resources/migrate/common/V_4_2_2_4__add_dlm_task_meta.sql [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+    `job_id` varchar(120) NOT NULL COMMENT 'job id for dlm task',
+    `processed_data_size` bigint(20) NOT NULL COMMENT 'the data size of task generator processed',
+    `processed_row_count` bigint(20) NOT NULL COMMENT 'number of rows processed by task generator',
+    `status` varchar(64) NOT NULL COMMENT 'status of the task generator',
+    `type` varchar(32) NOT NULL COMMENT 'type of the task generator',
+    `primary_key_save_point` varchar(512) DEFAULT NULL COMMENT 'save point for primary key',
+    `partition_save_point` varchar(512) DEFAULT NULL COMMENT 'save point for partition',
+    `task_count` bigint(20) NOT NULL DEFAULT '0' COMMENT 'number of tasks',
+    CONSTRAINT pk_dlm_task_generator_id PRIMARY KEY (id),
+    UNIQUE KEY `uk_dlm_task_generator_generatora_id` (`generator_id`),
```
Comment (yhilmare) [inline server/odc-migrate/src/main/resources/migrate/common/V_4_2_2_4__add_dlm_task_meta.sql:15]: type error: uk_dlm_task_generator_generator_id

**Review by yhilmare — APPROVED** [review]
lgtm

**Review by yhilmare — APPROVED** [review]
lgtm

**Pull request merged**